### ISSUE TITLE
All properties of client-s3 GetObjectCommandOutput described same as "Body"

### ISSUE THREAD

**Issue by IPCPaul** [aws/aws-sdk-js-v3#4111]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Since version 3.188.0 (last version without this bug 3.186.0) the client-s3 package describes all properties of GetObjectCommandOutput as "SdkStream<internal.Readable | ReadableStream | Blob | undefined> | undefined", unless the 'DOM' lib is included in the tsconfig.

I guess the change was introduced here:
https://github.com/aws/aws-sdk-js-v3/pull/3977/files
https://github.com/aws/aws-sdk-js-v3/commit/ad99b66944b6fe1ce83082c1a33193dff033cfaf

### SDK version number

@aws-sdk/client-s3@3.197.0, @aws-sdk/client-s3@3.186.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v16.18.0

### Reproduction Steps

```js
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3Client = new S3Client({});
const getObjectCommand = new GetObjectCommand({
  Bucket: "my-bucket-name",
  Key: "/my/s3/key"
});
(async () => {
  const r = await s3Client.send(getObjectCommand);
  const contentLength: number | undefined = r.ContentLength;
  console.log("GOT CONTENT LENGTH:", contentLength);
})()
  .then(() => {
    console.log("Fetched from s3");
  })
  .catch((error) => {
    console.log("error", error)
  })
  .finally(() => {
    console.log("OK, really done.")
  });
```

### Observed Behavior

The code doesn't compile where properties of GetObjectCommandOutput are expected to be of types other than "SdkStream<internal.Readable | ReadableStream | Blob | undefined> | undefined"

### Expected Behavior

The code compiles and properties of GetObjectCommandOutput can have types other than "SdkStream<internal.Readable | ReadableStream | Blob | undefined> | undefined"

### Possible Solution

Similar (in fact "almost" the same) to the suggestion I saw here https://github.com/aws/aws-sdk-js-v3/issues/3807 only use the ReadableStream when it is defined on the global scope.

Where the GetObjectOutput is defined (https://raw.githubusercontent.com/aws/aws-sdk-js-v3/main/clients/client-s3/src/models/models_0.ts) this seems to work:

```
type OptionalReadableStream = typeof globalThis extends { ReadableStream: infer GlobalReadableStream } ? GlobalReadableStream : never;
export interface GetObjectOutput {
  /**
   * <p>Object data.</p>
   */
  Body?: Readable | OptionalReadableStream | Blob;
  ... etc ...
  DeleteMarker?: boolean;
  ... etc ...
}
```

The OptionalReadableStream gets assigned to the global ReadableStream definition in case there is one. If there is none it becomes never.

### Additional Information/Context

With regards to this current work around:
```
declare global {
   interface ReadableStream {}
}
```

I "think" this is basically causing every property of the GetObjectOutput to be assignable to "Body" within this code:
```
export interface GetObjectCommandOutput extends __WithSdkStreamMixin<GetObjectOutput, "Body">, __MetadataBearer {
}
```

Please let me know if there's any more information I can provide.

**Comment by RanVaknin** [comment]
Hi @IPCPaul ,

> **The code doesn't compile** where properties of GetObjectCommandOutput are expected to be of types other than "SdkStream<internal.Readable | ReadableStream | Blob | undefined> | undefined"

How are you compiling? 

Are you running this locally using node?

Both myself and @yenfryherrerafeliz were not able to reproduce your issue.
Here is my repro

**_sample.ts:_**
```js
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3Client = new S3Client({});
const getObjectCommand = new GetObjectCommand({
  Bucket: "myBucket",
  Key: "hello.txt"
});
(async () => {
  const r = await s3Client.send(getObjectCommand);
  const contentLength: number | undefined = r.ContentLength;
  console.log("GOT CONTENT LENGTH:", contentLength);
})()
```

**_Compile ts to js and run:_**
```console
$ tsc sample.ts
$ node sample.js
GOT CONTENT LENGTH: 13
```

Not sure why you are running into issues, but if theres a specific error message can you share it with us?

Thanks,
Ran~

**Comment by Lewenhaupt** [comment]
I'm getting the same problem with @aws-sdk/client-s3@3.197.0 and @aws-sdk/types@3.197.0:
![image](https://user-images.githubusercontent.com/7331707/198531171-1d1bc759-80b6-4c8a-a046-e46189e943a9.png)

**Comment by Lewenhaupt** [comment]
Writing WithSdkStreamMixin as follows fixes the issue:
`export declare type WithSdkStreamMixin<T, StreamKey extends keyof T> = {
    [key in keyof T]: [T[key]] extends [T[StreamKey]] ? SdkStream<T[StreamKey]> : T[key];
};`

I believe this is the issue: https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types

**Comment by Lewenhaupt** [comment]
I took the liberty to provide a PR with the suggested changes.

**Comment by RanVaknin** [comment]
@Lewenhaupt ,

Thank you for looking this and providing a PR. I marked this to be reviewed by the dev team for further actions.

All the best,
Ran~

**Comment by kuhe** [comment]
I made a modification to the PR. We should target the key name for type transform, rather than attempting to match the type of the value at the key. I believe that will resolve the issue more conclusively. In my test code, wrapping in a length-1 array did not fix the issue.

**Comment by Lewenhaupt** [comment]
> I made a modification to the PR. We should target the key name for type transform, rather than attempting to match the type of the value at the key. I believe that will resolve the issue more conclusively. In my test code, wrapping in `[]` did not fix the issue. 

That's odd but I agree that your solution is better. However the docs for the function ought to be changed to since it won't change anything else than the key matching the streamKey.

**Comment by IPCPaul** [comment]
Thanks for everyone looking at this. Sorry for the slow reply.

My whole project looks like this:

**tsconfig.json**
```
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Node 16",

  "compilerOptions": {
    "lib": ["es2021"],
    "module": "commonjs",
    "target": "es2021",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  },
  "files": [
    "index.ts"
  ]
}
```

**package.json**
```
{
  "name": "test",
  "version": "1.0.0",
  "dependencies": {
    "@aws-sdk/client-s3": "3.197.0"
  },
  "scripts": {
    "compile": "tsc"
  }
}
```

**index.ts**
```
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";

const s3Client = new S3Client({});
const getObjectCommand = new GetObjectCommand({
  Bucket: "myBucket",
  Key: "hello.txt"
});
(async () => {
  const r = await s3Client.send(getObjectCommand);
  const contentLength: number | undefined = r.ContentLength;
  console.log("GOT CONTENT LENGTH:", contentLength);
})()
```

Then then to test it I'm running:
`tsc --project ./`

Which gives me:
```
index.ts:11:9 - error TS2322: Type 'SdkStream<Readable | ReadableStream | Blob | undefined> | undefined' is not assignable to type 'number | undefined'.
  Type 'Readable & SdkStreamMixin' is not assignable to type 'number | undefined'.
    Type 'Readable & SdkStreamMixin' is not assignable to type 'number'.

11   const contentLength: number | undefined = r.ContentLength;
```

The proposed PR does seem to fix it (https://github.com/aws/aws-sdk-js-v3/pull/4119)

**Comment by IPCPaul** [comment]
Fixed. The issue appears to be resolved in @aws-sdk/client-s3: 3.200.0 (by https://github.com/aws/aws-sdk-js-v3/pull/4119)
Thanks again everyone for helping with this.
